File: src/main/java/marcel/hardware/Sound.java
package marcel.hardware;

import javax.sound.sampled.*;
import javax.sound.sampled.AudioFormat.Encoding;
import javax.sound.sampled.DataLine.Info;

public class Sound {

    private final Thread player = null;

    private static final float frameRate = 44100f; // 44100 Hz
    private static final int channels = 1; // Mono
    private static final double duration = 0.020; // 20ms
    private static final int sampleBytes = Short.SIZE / 8; // 8 bits
    private static final int frameBytes = sampleBytes * channels;
    private static final AudioFormat format =
        new AudioFormat(Encoding.PCM_SIGNED,
                        frameRate,
                        Short.SIZE,
                        channels,
                        frameBytes,
                        frameRate,
                        true);
    private static final int nFrames = (int) Math.ceil(frameRate * duration);
    static int nSamples = nFrames * channels;
    private static final int nBytes = 1024; // Buffer size
    private static byte[] data;// = ByteBuffer.allocate(nBytes);
    static byte[] datain;

    private static SourceDataLine line;


    public Sound() {
        data = new byte[nBytes];
        // Réservation de la sortie audio, début de la restitution, envoi du tableau
        Info info = new Info(SourceDataLine.class, format);

        try {
            line = (SourceDataLine) AudioSystem.getLine(info);
            line.open(format);
        } catch (LineUnavailableException e) {
            e.printStackTrace();
        }
        line.start();

    }

    // Copie du buffer de son provenant du 6809 vers le buffer de la carte son
    // Cette fonction est lancée lorsque le buffer 6809 est plein

    public static synchronized void playSound(M6809 cpu) {
        for (int i = 0; i < nBytes; i++) {
            data[i / 4] = cpu.getSound_buffer(i);
        }
        line.write(data, 0, nBytes / 4);
    }
}

File: src/hardware/sound.rs
use std::time::Duration;

use rodio::{OutputStream, OutputStreamHandle, Sink, Source};
use rodio::source::SineWave;

use crate::hardware::M6809::M6809;
use crate::int;

const FRAME_RATE: u32 = 44100;
// 44100 Hz
const DURATION: Duration = Duration::from_millis(20);
const SAMPLE_BYTES: int = 16 / 8; // 8 bits

const N_FRAMES: int = 882;
const N_BYTES: usize = 1024; // Buffer size

pub(crate) struct Sound {
    stream: OutputStream,
    line: OutputStreamHandle,
}

impl Sound {
    pub(crate) fn new() -> Self {
        let (stream, line) = OutputStream::try_default().unwrap();
        Sound {
            stream,
            line,
        }
    }

    // Copie du buffer de son provenant du 6809 vers le buffer de la carte son
    // Cette fonction est lancée lorsque le buffer 6809 est plein
    pub(crate) fn play_sound(&mut self, cpu: &M6809) {
        let mut has_data = false;
        let mut data = vec![0f32; N_BYTES];
        for i in 0..N_BYTES {
            let value = cpu.sound_buffer[i];
            if value != 0 {
                has_data = true;
                data[i / 4] = value as f32 /  u8::MAX as f32;
            }
        }
        if has_data {
            let pcm = PcmSource {
                data,
                position: 0,
            };

            // let sine = SineWave::new(440.0);
            // self.line.play_raw(sine).ok();
            let samples_converter = pcm.convert_samples();
            self.line.play_raw(samples_converter).ok();
        }
    }
}

struct PcmSource {
    data: Vec<f32>,
    position: usize,
}

impl Iterator for PcmSource {
    type Item = f32;

    #[inline]
    fn next(&mut self) -> Option<Self::Item> {
        if self.position < self.data.len() {
            let result = self.data[self.position];
            self.position += 1;
            Some(result)
        } else {
            None
        }
    }
}

impl Source for PcmSource {
    fn current_frame_len(&self) -> Option<usize> {
        None
    }

    fn channels(&self) -> u16 {
        1
    }

    fn sample_rate(&self) -> u32 {
        FRAME_RATE
    }

    fn total_duration(&self) -> Option<Duration> {
        None
    }
}